Specify data collection (#59)
abetterinternet/prio-documents: draft-pda-protocol.md

   input-validation protocol.
1. Report: Uploaded to the leader from the client. A report contains the

>    secret-shared and encrypted input and proof.

Reviewer: Since we're getting into the -arity of things: does a report contain shares of a _single_ input or could several inputs be encoded into a single report?

Author: I think one input per report. As has been said, it might be that the "input" is actually comprised of multiple measurements. The "input" is meant to capture the thing that's validated in a single run of the input validation protocol.

## Data flow

Each PA task consists of two sub-protocols, *upload* and *collect*, which are
executed concurrently. Each sub-protocol consists of a sequence of HTTP requests
made from one entity to another.

~~~~
                   +------------+
                   |            |
    +-------------->   Helper   |
    |              |            |
    |              +-----^------+
    |                    |
    |                    |

>     | 1.                 |

Author: The leader doesn't interact with the helper during upload, only during collect.

made from one entity to another.

~~~~
+-------------+ 1.      +-------------+
|             +--------->             |
|   Client    |         |   Leader    |
|             |    +----+             |
+------+------+    | 2. +------^------+
       | 1.        |           |
       |           |           |
       |           |           | 2.
+------v------+    |    +------+------+
|             <----+    |             |
|   Helper    |         |  Collector  |
|             |         |             |
+-------------+         +-------------+
~~~~
{: #pa-topology title="Who makes requests to whom while executing a PA task."}

1. **Upload:** Each client assembles its measurements into an input for the
   given PA protocol. It generates a proof of its input's validity and splits
   the input and proof into two shares, one for the leader and another for a
   helper. Rather than send each share to each aggregator directly, the client
   sends a request to the helper to learn its public key, encrypts its share
   under this public key, and sends the ciphertext to the leader. The leader's
   share and the helper's encrypted share are referred to collectively as the
   client's *report*.

> 2. **Collect:** The collector makes a request to the leader in order to obtain the

Reviewer 2: ```suggestion
2. **Collect:** The collector makes one or more requests to the leader in order to obtain the
```

   final output of the protocol. Before the output can be computed, the
   aggregators (i.e, the leader and helpers) need to have verified and
   aggregated a sufficient number of inputs. Depending on the PA protocol, it
   may be possible for the aggregators to do so immediately when reports are
   uploaded. (See {{prio}}.) However, in general it is necessary for them to
   wait until (a) enough reports have been uploaded and (b) the collector has
   made a request. (See {{hits}}.)

[TODO: Say that the protocol involves multiple helpers. The downside of using
secret sharing is that the protocol requires at least two servers to be online
during the entire data aggregation process. To ameliorate this problem, we run

> the protocol in parallel with multiple helpers.]

Reviewer 2: Can we move this to an issue and discuss separately? What does "we run the protocol in parallel with multiple helpers" actually mean? (Who is "we"?)

Author: We decided previously that multiple helpers would be used for resiliency. This has already been baked into the spec. This TODO is merely about motivating and describing this design choice. There are two issues that are impacted by this: #4 and #22.

Reviewer 2: I'll raise as a separate issue as I no longer thing this is necessary to bake into the protocol this way.

Author: Done -> https://github.com/abetterinternet/prio-documents/issues/68

## Configuration {#pa-config}

### Tasks

Each PA protocol is associated with a *PA task* that specifies the measurements
that are to be collected:

~~~
struct {
  uint16 version;
  opaque id[16];
} PATask;
~~~

The first field, `version` specifies the version of this document. The second
field, `id`, is an opaque identifier used by the clients, aggregators, and
collector to uniquely identify the PA task at hand. We will call it the *task
id* in the remainder.

[TODO: Decide how the `PATask` is configured. Eventually this will be
distributed, in an authenticated manner, from the collector to the other
parties. For now, we just assume this value is negotiated out-of-band.]

### Parameters

Associated to each task are the *PA Parameters* encoded by the `PAParam`
structure:

~~~
struct {
  PATask task;
  Url helper_urls<1..2^16-1>;

>   HpkeConfig collector_config;

Author: Yeah, some spelling of this is a good idea I think. I've left this as an OPEN ISSUE for now, since it's somewhat orthogonal. See below.

} PAParam;

enum { prio(0), hits(1) } PAProto;

opaque Url<1..2^16-1>;
~~~

* `task`: The PA task.
* `helper_urls`: The helpers' endpoint URLs.
* `collector_config`: The HPKE configuration of the collector (described in
  {{hpke-config}}). [OPEN ISSUE: Maybe the collector's HPKE config should be
  carried by the collect request?]
* `batch_size`: The batch size, i.e., the minimum number of reports that are
  aggregated into an output.

> * `batch_window`: The batch window, i.e., the minimum time difference between

Reviewer 2: Are we concerned about skew between collectors and aggregators?

Reviewer 2: We should then (a) note that as a requirement and (b) note what happens if that does not happen.

Author: More precisely, I think all that's required for privacy is that each aggregator needs to have a clock that's roughly in sync with true time (within batch_window seconds). Added this to "Pre-conditions".

opaque Url<1..2^16-1>;
~~~

* `task`: The PA task.
* `helper_urls`: The helpers' endpoint URLs.
* `collector_config`: The HPKE configuration of the collector (described in
  {{hpke-config}}). [OPEN ISSUE: Maybe the collector's HPKE config should be

>   carried by the collect request?]

Reviewer: This would allow the leader to impersonate the collector when it relays the HPKE config to the helper -- but as you say, it's an open issue so we don't have to solve it right now.

Author: Actually, at the moment the helper is "pre-configured" with the PAParam (and thus the collectors HPKE config). But yeah, this something we have to work out.

uint16 HpkeKemId;  // Defined in I-D.irtf-cfrg-hpke
uint16 HpkeKdfId;  // Defined in I-D.irtf-cfrg-hpke
~~~
[TODO: Decide whether to use the same config structure as OHTTP/ECH. This would
add support for multiple cipher suites.]

We call this a *key configation*. The key configuration is used to set up a
base-mode HPKE context to use to derive symmetric keys for protecting: (1) input
shares sent from the client to the helper; or (2) output shares sent from the
helper to the collector. The *config id*, `HpkeConfig.id`, is forwarded by the
sender to the receiver to help the receiver decide if it knows the decryption
key.

## Pre-conditions

We assume the following conditions hold before execution of any PA task begins:

1. The aggregators agree on a set of PA tasks, as well as the PA parameters
   associated to each task.

> 1. Each client is configured with a specific PA task.

Author: Yes indeed. I don't this criterion is very clear. What we mean is "each client has selected a PA task for which it will upload a report". Fixed.

1. Each client knows the URL of the leader endpoint, e.g.,
   `example.com/metrics`. We write this URL as `[leader]` below. (We write
   `[helper]` for a helper's URL.)
1. Each client and the leader can establish a leader-authenticated secure
   channel.
1. The leader and each helper can establish a helper-authenticated secure
   channel.
1. The collector and leader can establish a leader-authenticated secure channel.
1. The collector has chosen an HPKE key pair.
1. Each helper has chosen an HPKE key pair.
1. Each helper has chosen a key for an AEAD scheme.

[TODO: It would be clearer to include a "pre-conditions" section prior to each
"phase" of the protocol.]

## Upload {#pa-upload}

~~~~
Client            Leader
  |  upload start   |
  <----------------->
  |  upload finish  |
  <----------------->

>   v                 v

Author: Good call. Done.

~~~
struct {
  PAProto proto;
  select (PALeaderInputShare.proto) {
    case prio: PrioLeaderInputShare;
    case hits: HitsLeaderInputShare;
  };
} PALeaderInputShare;
~~~

Note that only the helper's share is encrypted, and that the leader share is

> sent in plaintext.

Author: Noted.

All we say about transport is that it should be a "server-authenticated secure channel". (I don't think we should be more specific than this.)

and L aggregate requests per collect request."}

[TODO: Decide how the collector's request is authenticated.]

The collector interacts with the leader to produce the final aggregate output.
This process consists of a sequence of *collect requests* issued to the leader.
Before a request can succeed, the aggregators must have verified and aggregated
enough reports and the leader must have obtained the helper's encrypted output
share (see {{pa-aggregate}}). In general, the procedure by which the aggregators
verify and aggregate reports depends on parameters carried by the collect
request. This procedure is described below in {{pa-aggregate}}.

### Collect Request

The collect protocol is an iterative procedure driven by the collector and
parameterized by a PAParam. At a high level, it proceeds as follows. First, the
collector initializes local per-protocol state using the corresponding
PAParam. This state object has the following member functions:

- CreateRequest() -> PACollectReq: Creates a PACollectReq object, defined below,
  that is sent to the leader to complete one iteration of the protocol.

~~~
struct {
  PATask task;

>   Url helper;

Author: > A `PAParam` can contain _n_ `helper_urls`, but a client may only select _k_ < _n_ of them to send reports to. So how does the collector know which `helper_urls` to use? I'm guessing that collectors are meant to send a `PACollectReq` for all _n_ helpers, but if _k_ << _n_ this seems wasteful. Are we assuming that _k_ is exactly or close to _n_?

The collector knows `helper_urls` because it's now specified by `PAParam`. (Note that this is change is new to this PR.) Nominally, _k_ would be close to _n_ (and _n_ would be be around 2 or 3).

> Supposing that some helper did not participate, what does the `PACollectResp` look like? Does it return an error (which should be described in 3.x. Common abort conditions) or does it provide a `PACollectResp` with empty `leader_share` and `encrypted_helper_share`?

Good question. (This is subsumed by the OPEN ISSUE below; I added your particular question as an example.) I think the right answer is that the leader would "abort" the request as described in {{pa-error}}. An important thing to note about this case is that it's (partially) recoverable, since the collector can try again with another helper.

> What about the case where `PACollectReq.helper` does not occur in `PAParam.helper_urls`? Does that get rejected by leader as an invalid message, or treated the same as a valid helper that did not participate?

This would be another "abort" I think.

>   Url helper;

Reviewer 2: It seems a lot of complexity right now is coming from the fact that the collector allows for multiple helpers to be used by the client, and the client to make a selection. Can we just remove that option, and have the collector force which helpers to use?

>   Url helper;

Author: This discussion seems regressive given that we've already decided that multiple helpers would be used to add resiliency. If we want to discuss changing course, then let's do so in an separate issue.

>   Url helper;

Author: Filed: https://github.com/abetterinternet/prio-documents/issues/68

>   PAProto proto;

Author: But there is no `PAParam` in this struct. I guess you could argue that the leader knows the `PAParam` since it has the task id. However, I tend to think that protocol messages should always be deserializaeable without extra context.

  select (PACollectReq.proto) {
    case prio: PrioCollectReq;
    case hits: HitsCollectReq;
  }
} PACollectReq;
~~~

- Update(resp: PACollectResp): Consumes an opaque PACollectResp object, defined
  below, that is received from the leader in response to a PACollectReq.

~~~
struct {
  PATask task;

>   PAProto proto;

Author: I see what you mean. As I mentioned above, I think protocol messages should always be deserializeable without needing context. Another way of phrasing this is that protocol messages should be "self describing". Maybe this property isn't needed here.

  PAOutputShare leader_share;

>   opaque encrypted_helper_share;

Author: The purpose of encryption here is to ensure that the aggregators don't learn the final output. (This is one of the requirements enumerated in #51.) This is a useful property to have if the collector "outsources" aggregation to a third-party leader. (See @ekr's comment on #51.)

>   opaque encrypted_helper_share;

Author: It also occurs to me that encrypting the output share is important to prevent the leader from doing Sybil attacks.

return state.Output()
~~~

[OPEN ISSUE: Describe how intra-protocol errors yield collect errors (see
issue#57).]

Each collect request involves one of the helpers specified by the PA parameters.
If more than one helper is specified, the collector may issue the requests in
any order.

### Verifying and Aggregating Reports {#pa-aggregate}

[TODO: Specify how the helper encrypts its state.]

After the client uploads a report to the leader, the leader and helper verify in
zero knowledge that the proof is well-formed. The exact procedure for doing so
is protocol specific, but all protocols have the same basic structure. In
particular, the protocol is comprised of a sequence of *aggregate requests* from
the leader to the helper. At the end of this procedure, the leader and helper
will have have aggregated a set of valid client inputs.

#### Aggregate Request

The process begins with a PACollectReq. The leader collects a sequence of
reports that are all associated with the same PA task, helper URL, and helper
HPKE config id. Let `[helper]` denote `PACollectReq.helper_url`. The leader
sends a POST request to `[helper]/aggregate` with the following message:

~~~
struct {
  PATask task;

>   uint8 helper_hpke_config_id;

Author: The config_id is meant to uniquely identify the HPKE config. We're using the same type for the config_id in ODoH and ECH, and it hasn't been a problem there. Usually the server will only have a small number of keys at any one time, e.g., 3 or 4.

>   uint8 helper_hpke_config_id;

Reviewer: I had assumed that a helper wouldn't re-use HPKE configs across PATasks, but thinking about it further there's no reason to assume that.

The leader responds to well-formed requests to `[leader]/upload_finish` with
status 200 and an empty body. Malformed requests are handled as described in
{{pa-error-common-aborts}}.

## Collect {#pa-collect}

~~~~
Collector     Leader           Helper
  |  collect 1  |                |
  +------------->                |
  |             |  aggregate 1   |
  |             <---------------->
  |             |  ...           |
  |             |  aggregate L   |
  |             <---------------->
  |             |  output share  |
  |             <---------------->
  <-------------+                |
  |  ...        |                |
  |  collect N  |                |
  +------------->                |
  |             |  aggregate 1   |
  |             <---------------->
  |             |     ...        |
  |             |  aggregate L   |
  |             <---------------->
  |             |  output share  |
  |             <---------------->
  <-------------+                |
  v             v                v
~~~~
{: #pa-collect-flow title="Flow of the collect process with N collect requests
and L aggregate requests per collect request."}

> [TODO: Decide how the collector's request is authenticated.]

Reviewer 2: ```suggestion
[TODO: Decide if and how the collector's request is authenticated.]
```

Reviewer 2: Why does a collector's request to get the aggregate output need to be authenticated? That seems like a deployment specific property.

Author: I don't think this is going to end up being deployment specific, but I'll concede the point for now.

After the client uploads a report to the leader, the leader and helper verify in
zero knowledge that the input is valid. The exact procedure for doing so is
protocol specific, but all protocols have the same basic structure. In
particular, the protocol is comprised of a sequence of *aggregate requests* from
the leader to the helper. At the end of this procedure, the leader and helper
will have have aggregated a set of valid client inputs.

#### Aggregate Request

The process begins with a PACollectReq. The leader collects a sequence of
reports that are all associated with the same PA task, helper URL, and helper
HPKE config id. Let `[helper]` denote `PACollectReq.helper_url`. The leader
sends a POST request to `[helper]/aggregate` with the following message:

~~~
struct {
  PATask task;
  uint8 helper_hpke_config_id;
  opaque helper_state<0..2^16>;
  PAAggregateSubReq seq<1..2^24-1>;
} PAAggregateReq;
~~~

The structure contains the PA task, the helper's HPKE config id, an opaque
*helper state* string, and a sequence of *sub-requests*, each corresponding to a
unique client report.  Sub-requests are structured as follows:

~~~
struct {
  opaque enc<0..2^16-1>;

>   uint64 time; // UNIX timestamp (seconds) of the report.

Reviewer 2: This is generated by the leader, presumably? Is it set when the report is received from the client, or when aggregation is actually done during collection? Could the client set the timestamp in its encrypted share for the helper?

Reviewer 2: What stops the leader from altering the timestamp, then? Should it be folded into the helper AAD, perhaps?

Author: Yup, that's the idea. There's a TODO for this in "Upload Finish Request".

  opaque helper_state<0..2^16>;
  PAAggregateSubResp seq<1..2^24-1>;
} PAAggregateResp;

struct {
  PAProto proto;
  select (PAAggregateSubResp.proto) {
    case prio: PrioAggregateSubResp;
    case hits: HitsAggregateSubResp;
  }
} PAAggregateSubResp;
~~~

The helper handles each sub-request `PAAggregateSubReq` as follows.  It first
checks that checks that `PAAggregateSubReq.proto == PAParam.proto`. If not, it
aborts and alerts the leader with "incorrect protocol for sub-request".
Otherwise, It computes the HPKE context as

~~~
context = SetupBaseR(PAAggregateSubReq.enc, sk, [TODO])
~~~

where `sk` is its HPKE key and computes the body of the `PAAggregateSubResp`. It
then updates its state according to the PA protocol. After processing all of the
sub-requests, the helper encrypts its updated state and constructs its response
to the aggregate request.

##### Helper State

The helper state is an optional parameter of an aggregate request that the can
helper use to carry state accross requests. At least part of the state will
usually need to be encrypted in order to protect user privacy. However, the
details of precisely how the state is encrypted and the information that it

> carries is up to the implementation of the helper.

Reviewer 2: ```suggestion
carries is up to the helper implementation.
```

#### Output Share Request

Once the aggregators have verified at least as many reports as required for the
PA task, the leader issues an *output share request* to the helper. The helper
responds to this request by extracting its output share from its state and
encrypting it under the collector's HPKE public key.

The leader sends a POST request to `[helper]/output_share` with the following
message:

~~~
struct {
  PATask task;
  opaque helper_state<0..2^16>;

> } PAOutputShareReq;

Reviewer 2: Hmm, where does the batch window come into play here? Shouldn't it be the case that aggregate outputs are only produced for a given batch window once the size is hit? Do the collector's request parameters need to fold into this OutputShareReq?

Author: Ah, good catch. I added the `batch_start` and `batch_end` parameters so that the helper can enforce the `batch_window` and `batch_size` parameters.

~~~

To respond to valid output share requests, the helper begins by decrypting its

> state. It then constructs its output share, which has the following structure:

Reviewer 2: Should we require that the helper check that the OutputShareReq is not sent before the batch limit is hit? (That is, should the helper track how many shares aggregated so far and error if the number is less than the size?)

Author: Yes indeed. I updated the text to make this check explicit.

  |             |  aggregate L   |
  |             <---------------->
  |             |  output share  |
  |             <---------------->
  <-------------+                |
  v             v                v
~~~~
{: #pa-collect-flow title="Flow of the collect process with N collect requests
and L aggregate requests per collect request."}

[TODO: Decide if and how the collector's request is authenticated.]

The collector interacts with the leader to produce the final aggregate output.
This process consists of a sequence of *collect requests* issued to the leader.
Before a request can succeed, the aggregators must have verified and aggregated
enough reports and the leader must have obtained the helper's encrypted output
share (see {{pa-aggregate}}). In general, the procedure by which the aggregators
verify and aggregate reports depends on parameters carried by the collect
request. This procedure is described below in {{pa-aggregate}}.

### Collect Request

The collect protocol is an iterative procedure driven by the collector and
parameterized by a PAParam. At a high level, it proceeds as follows. First, the
collector initializes local per-protocol state using the corresponding

> PAParam. This state object has the following member functions:

Reviewer 3: I'd prefer we don't describe this in this object oriented fashion. It's incredibly hard to read. Just talk about the protocol messages, not member functions.

Author: This was @chris-wood's idea. I think it's nice, but not strictly necessary. I added a TODO to decide whether to change this.

Reviewer 3: OK, I think it's a regression.

- CreateRequest(): Creates a PACollectReq object, defined below, that is sent to
  the leader to complete one iteration of the protocol. Along with any
  protocol-specific parameters, the message specifies a time interval
  `[batch_start, batch_end)` that determines the batch of reports to be
  aggregated. It must be that `batch_end - batch_start >= PAParam.batch_window`
  and `batch_start` and `batch_end` are multiples of `PAParam.batch_window`.

~~~
struct {
  PATask task;
  Url helper;
  uint64 batch_start; // The beginning of the batch in UNIX time.
  Uint64 batch_end;   // The end of the batch in UNIX time (exclusive).
  PAProto proto;
  select (PACollectReq.proto) {
    case prio: PrioCollectReq;

>     case hits: HitsCollectReq;

Reviewer 3: Do we understand what the contents of this might be? Are they really going to be per-request? In particular, why do we need ```proto```. Each task should just have one proto

Author: - PrioCollectReq would minimally have some identifier for the validity circuit and, potentially, the field.
- HitsCollectReq might might specify one or two fields.

Reviewer 3: I don't see why any of this needs to be per-request. They should just be fixed per-task.

Author: For Hits this is also going to convey the set of candidate prefixes. It's therefore not fixed per request.

Reviewer 3: Well, we should definitely factor out the things that are per-task and things that are per-collector-request.

  }
} PACollectReq;
~~~

- Update(resp: PACollectResp): Consumes an opaque PACollectResp object, defined
  below, that is received from the leader in response to a PACollectReq.

~~~
struct {
  PATask task;
  PAProto proto;
  PAOutputShare leader_share;
  opaque encrypted_helper_share;
} PACollectResp;
~~~

- Finished(): Returns a boolean indicating indicating whether or not the collection
  procedure is complete. This function returns true when the collect iteration
  is complete.
- Output(): Returns the aggregate output corresponding to the protocol.

The collect procedure for a given PAParam structure `param` is then driven with
the following algorithm:

~~~
state = CreateState(param)
while not state.Finished():
   req = state.CreateRequest()
   resp = fetch([leader]/collect, req)
   state.Update(resp)
return state.Output()
~~~

[OPEN ISSUE: Describe how intra-protocol errors yield collect errors (see
issue#57). For example, how does a leader respond to a collect request if the
helper drops out?]

Each collect request involves one of the helpers specified by the PA parameters.
If more than one helper is specified, the collector may issue the requests in

> any order.

Reviewer 3: As noted in in issue #68 this just seems like extra complexity. The helpers for any given task should be fixed.

Author: Ack, I'm sending a PR today to rip this out.

### Verifying and Aggregating Reports {#pa-aggregate}

After the client uploads a report to the leader, the leader and helper verify in
zero knowledge that the input is valid. The exact procedure for doing so is
protocol specific, but all protocols have the same basic structure. In
particular, the protocol is comprised of a sequence of *aggregate requests* from
the leader to the helper. At the end of this procedure, the leader and helper
will have have aggregated a set of valid client inputs.

#### Aggregate Request

The process begins with a PACollectReq. The leader collects a sequence of
reports that are all associated with the same PA task, helper URL, and helper
HPKE config id. Let `[helper]` denote `PACollectReq.helper_url`. The leader
sends a POST request to `[helper]/aggregate` with the following message:

~~~
struct {
  PATask task;
  uint8 helper_hpke_config_id;
  opaque helper_state<0..2^16>;
  PAAggregateSubReq seq<1..2^24-1>;
} PAAggregateReq;
~~~

> The structure contains the PA task, the helper's HPKE config id, an opaque

Reviewer 3: What is the hpke_config_id used for? And why is it only one octet long?

Author: [Oops, I thought I commented on this, but somehow the comment didn't make into the queue.] We're following the precedent set by ECH here: the config id is just a byte. It's used by the decrypting party to figure out which key to use.

*helper state* string, and a sequence of *sub-requests*, each corresponding to a
unique client report.  Sub-requests are structured as follows:

~~~
struct {
  opaque enc<0..2^16-1>;
  uint64 time; // UNIX timestamp (seconds) of the report.

>   PAProto proto;

Reviewer 3: Again, why is ```proto``` being carried here? You can't change it on the fly.

Author: The structure of the rest of this message depends on the PA protocol. Maybe there's a better way to express this?

Reviewer 3: You don't need to carry that in the message. You just do like ```select (proto_in_use)``` or something and then note in the text that it's pointing to something external.

Author: Added a TODO to address this later.

      PrioAggregateSubReq;
    case hits:
      HitsHelperInputShare;
      HitsAggregateSubReq;
  }
} PAAggregateSubReq;
~~~

The `enc` field is the helper's encapsulated HPKE context sent in the report.
(This field is marked optional because it is usually only needed for the first
aggregate request. [TODO: Decide how to mark the helper input share as optional
as well. This too is usually only needed for the first request.]) The `time`
field should match the timestamp of the corresponding report. The remainder of
the structure is dedicated to the protocol-specific helper share and request
parameters used for the current round.

The helper handles well-formed requests as follows. (As usual, malformed
requests are handled as described in {{pa-error-common-aborts}}.) It first looks
for the PA parameters `PAParam` for which `PAAggregateReq.task.id ==
PAParam.task.id`. Next, it looks up the HPKE config and corresponding secret key
associated with `PAAggregateReq.helper_hpke_config_id`. If not found, then it
aborts and alerts the leader with "unrecognized key config". [NOTE: In this
situation, the leader has no choice but to abort. This falls into the class of

> error scenarios that are addressable by running with multiple helpers.]

Reviewer 3: Why are you assuming that each report uses the same key_id. Suppose you had a situation in which you had a batch of a day's reports and the helper changed keys halfway through?

Author: In that case, the leader would need to split the reports into two requests.

> Why are you assuming that each report uses the same key_id.

It seemed to me like it would be simpler if there's just one HPKE key used per request. 

> Suppose you had a situation in which you had a batch of a day's reports and the helper changed keys halfway through?

The leader would split the reports into two separate requests, one for the old key and another for the new.

Reviewer 3: But that might end up with a tail that could never be aggregated because it was smaller than the minimal batch size.

Author: Not true, since the number of "sub-requests" isn't necessarily the same as the batch size. The leader makes several aggregate requests and not just one.

In any case, I'm good with relaxing this constraint. Added a TODO.

The response consists of the helper's updated state and a sequence of
*sub-responses*, where the i-th sub-response corresponds to the sub-request for
each i. The structure of each sub-response is specific to the PA protocol:

~~~
struct {
  opaque helper_state<0..2^16>;
  PAAggregateSubResp seq<1..2^24-1>;
} PAAggregateResp;

struct {
  PAProto proto;
  select (PAAggregateSubResp.proto) {
    case prio: PrioAggregateSubResp;
    case hits: HitsAggregateSubResp;
  }
} PAAggregateSubResp;
~~~

The helper handles each sub-request `PAAggregateSubReq` as follows.  It first
checks that checks that `PAAggregateSubReq.proto == PAParam.proto`. If not, it

> aborts and alerts the leader with "incorrect protocol for sub-request".

Reviewer 3: This could be avoided by just not having these fields.

Author: Ack, added a TODO.

  uint64 batch_start; // Same as PACollectReq.batch_start.
  Uint64 batch_end;   // Same as PACollectReq.batch_end.
  opaque helper_state<0..2^16>;
} PAOutputShareReq;
~~~

To respond to valid output share requests, the helper first checks that
`batch_start` and `batch_end` are multiples of `task.batch_window` and that
`batch_end - batch_start >= task.batch_window`. Next, it extracts from its state
the set of input shares that fall in the window `[batch_start, batch_end)`. If
the size of the batch is less than `task.batch_size`, then it aborts and alerts
the leader with "insufficient data". Otherwise, it computes its output share,
which has the following structure:

~~~
struct {
  PAProto proto;
  select (PAOutputShare.proto) {
    case prio: PrioOutputShare;
    case hits: HitsOutputShare;
  }
} PAOutputShare;
~~~

Next, it encrypts its output share under the collector's HPKE public key:

~~~
enc, context = SetupBaseS(pk, [TODO])
encrypted_output_share = context.Seal(output_share, [TODO])
~~~

where `pk` is the HPKE public key encoded by the collector's HPKE key
configuration and `output_share` is its serialized output share.

It responds with the following message:

~~~
struct {

>   uint8 collector_hpke_config_id;

Reviewer 3: I don't see how this works and why is it just one byte?

Author: This is basically the same thing we do in ECH: there is a 1-byte config identifier that's used to inform the decrypting party which HPKE key it needs to decrypt. It's a byte only because other HPKE applications have only a byte for the config id.

collector initializes local per-protocol state using the corresponding
PAParam. This state object has the following member functions:

- CreateRequest(): Creates a PACollectReq object, defined below, that is sent to
  the leader to complete one iteration of the protocol. Along with any
  protocol-specific parameters, the message specifies a time interval
  `[batch_start, batch_end)` that determines the batch of reports to be
  aggregated. It must be that `batch_end - batch_start >= PAParam.batch_window`

>   and `batch_start` and `batch_end` are multiples of `PAParam.batch_window`.

Reviewer: I like making the collector specify the batch window, as I think it absolves us of specifying some tricky edge cases w.r.t. insufficient # of reports in the window (we can just say that it's up to the collector to try again with wider batch windows until they find one that works).

Reviewer 2: Or maybe the collector can try again with a smaller window if it wants a different slice?
